llvm/lib/Target/AIE/AIERegClassConstrainer.cpp

```diff
@@ -0,0 +1,135 @@
+//===-- AIERegClassConstrainer.cpp - Constrain Reg classes before RA ------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineRegisterInfo.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <map>
+
+#include "AIE.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-ra"
+
+namespace {
+
+/// Constrain partially allocatable classes to fully allocatable ones.
+/// This makes sure that greedy works with classes for which it is easy to
+/// generate spill code.
+class AIERegClassConstrainer : public MachineFunctionPass {
+
+public:
+  static char ID;
+  AIERegClassConstrainer() : MachineFunctionPass(ID) {}
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addPreserved<SlotIndexes>();
+    AU.addPreserved<LiveIntervals>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+  /// Find the largest subclass of \p RC that is suitable for register
+  /// allocation. That is, it does not contain reserved registers, as those
+  /// are harder to spill/reload.
+  const TargetRegisterClass &getSuitableRCForRA(const TargetRegisterClass &RC,
+                                                const MachineRegisterInfo &MRI);
+
+  /// Cache for \p getSuitableRCForRA
+  std::map<unsigned, const TargetRegisterClass *> SuitableRCForRA;
+};
+
+bool AIERegClassConstrainer::runOnMachineFunction(MachineFunction &MF) {
+  LLVM_DEBUG(llvm::dbgs() << "Constraining RCs for RA: " << MF.getName()
+                          << "\n");
+  MachineRegisterInfo &MRI = MF.getRegInfo();
+  const TargetRegisterInfo &TRI = *MRI.getTargetRegisterInfo();
+
+  // Clear previous cache, reserved registers can change for each function
+  SuitableRCForRA.clear();
```

Comment:
I can suggest that we don't need to clean this cache always (from the compile time perspective). You can encapsulate this map with a BitVector representing the current set or reserved regs in a standalone class/struct (you can access the reserved registers with `BitVector getReservedRegs`  from TRI).   

We can have, in the end something like:

`SuitableRCForRA.clearIfDiverges(TRI)` -> just clean if we have different sets here.

Reply:
Done